iceoryx_examples/waitset/ice_waitset_trigger.cpp

```diff
@@ -0,0 +1,228 @@
+// Copyright (c) 2020 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iceoryx_posh/popo/wait_set.hpp"
+#include "iceoryx_posh/runtime/posh_runtime.hpp"
+#include "iceoryx_utils/cxx/optional.hpp"
+
+#include <iostream>
+#include <thread>
+
+// The two events the MyTriggerClass offers
+enum class MyTriggerClassEvents
+{
+    PERFORMED_ACTION,
+    ACTIVATE
+};
+
+// Triggerable class which has two events an both events can be
+// attached to a WaitSet.
+class MyTriggerClass
+{
+  public:
+    MyTriggerClass() = default;
+    ~MyTriggerClass() = default;
+
+    // IMPORTANT: For now the WaitSet does not support that the origin is moved
+    //            or copied. To support that we have to inform the waitset about
+    //            our new origin, otherwise the WaitSet would end up in the wrong
+    //            memory location when it calls the `hasTriggerCallback` with the
+    //            old origin (already moved) origin pointer. The same goes for
+    //            the resetCallback which is used when the WaitSet goes out of scope
+    //            and is pointing also to the old origin.
+    MyTriggerClass(const MyTriggerClass&) = delete;
+    MyTriggerClass(MyTriggerClass&&) = delete;
+    MyTriggerClass& operator=(const MyTriggerClass&) = delete;
+    MyTriggerClass& operator=(MyTriggerClass&&) = delete;
+
+    // When you call this method you will trigger the ACTIVATE event
+    void activate(const int activationCode) noexcept
+    {
+        m_activationCode = activationCode;
+        m_isActivated = true;
+        m_activateTrigger.trigger();
+    }
+
+    // Calling this method will trigger the PERFORMED_ACTION event
+    void performAction() noexcept
+    {
+        m_hasPerformedAction = true;
+        m_actionTrigger.trigger();
+    }
+
+    int getActivationCode() const noexcept
+    {
+        return m_activationCode;
+    }
+
+    // required by the m_actionTrigger to ask the class if it was triggered
+    bool hasPerformedAction() const noexcept
+    {
+        return m_hasPerformedAction;
+    }
+
+    // required by the m_activateTrigger to ask the class if it was triggered
+    bool isActivated() const noexcept
+    {
+        return m_isActivated;
+    }
+
+    // reset PERFORMED_ACTION and ACTIVATE event
+    void reset() noexcept
```

Comment:
Is this cool? When One event wakes up the waitset und you reset both conditions, you will loose some, or?

Reply:
You are right. It is fixed.